design/ip-filtering-design.md

```diff
@@ -0,0 +1,140 @@
+# IP Filtering 
+
+Status: Draft
+
+## Abstract
+
+Allow or deny traffic to services behind contour with an IP allowlist or an IP denylist.
+
+## Background
+
+There have been [long standing requests](https://github.com/projectcontour/contour/issues/62) to support some form of IP filtering.
+IP filtering is a useful tool for network segregation that is supported by other ingress controllers like [ingress-nginx](https://github.com/kubernetes/ingress-nginx/blob/87d1b8bbf28265386b339e65d8943d8c3f8582ed/docs/user-guide/nginx-configuration/annotations.md#whitelist-source-range) and [ambassador](https://www.getambassador.io/docs/edge-stack/latest/topics/running/ambassador#ip-allow-and-deny), and is a [common](https://github.com/projectcontour/contour/issues/62#issuecomment-698141472) [request](https://github.com/projectcontour/contour/issues/62#issuecomment-843193956) among the community, often listed as a [blocker](https://github.com/projectcontour/contour/issues/3693#issuecomment-1317049708) for adopting contour.
+
+## Goals
+
+- Support ip allowlists and denylists via Contour APIs.
+- Support filtering on request IP or on forwarded IP (i.e. via `X-Forwarded-For`)
+
+## Non Goals
+- Define Gateway API support for ip filtering. This is an obvious next step, but requires broader agreement.
+
+## High-Level Design
+
+The `route` section of an `HTTPProxy` will have two new fields added: `ipAllowPolicy` and `ipDenyPolicy`.
+
+Example `HTTPProxy`:
+
+```yaml
+apiVersion: projectcontour.io/v1
+kind: HTTPProxy
+metadata:
+  name: basic
+spec:
+  virtualhost:
+    fqdn: foo-basic.bar.com
+  routes:
+    - conditions:
+      - prefix: /
+      services:
+        - name: s1
+          port: 80
+      ipAllowPolicy:
+      # traffic is allowed if it came from localhost (i.e. co-located reverse proxy)
+      - peer: 127.0.0.1
+      # and the request originated from an IP in this range 
+      - remote: 99.99.0.0/16
+```
+
+## Detailed Design
+
+
+### API
+
+`ipAllowPolicy`/`ipDenyPolicy` has the following structure:
+
+```yaml
+ipAllowPolicy:
+- peer: 127.0.0.1
+- remote: 99.99.0.0/16
+```
+
+```yaml
+ipDenyPolicy:
+- peer: 127.0.0.1 
+- remote: 99.99.0.0/16
+```
+
+Note that:
+- Allow rules cannot be mixed with Deny rules: it won't be evident if a deny rule should be an exception to an allow rule, or if an allow rule should be an exception to a deny rule.
+- Multiple allow/deny CIDR ranges can be specified by repeating a key.
+- `peer` refers to the request IP
```

Comment:
writing down how exactly these concepts fold into the Envoy RBAC configuration fields would be good to make sure to head off any confusion early

Reply:
Yep, I took that from Ambassador (it was linked in one of the original GH issues on ip filtering). I don't have a strong opinion on the fields names, if `direct_remote` and `remote` sound better.

Reply:
> Is the terminology `peer` and `remote` originating from Ambassador? Envoy seems to use [`direct_remote_ip`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#envoy-v3-api-field-config-rbac-v3-principal-direct-remote-ip) and [`remote_ip`](https://www.envoyproxy.io/docs/envoy/latest/api-v3/config/rbac/v3/rbac.proto#envoy-v3-api-field-config-rbac-v3-principal-remote-ip). Which one is better to align?

I like the latter